Android/app/src/main/java/app/intra/sys/NetworkManager.java

```diff
@@ -113,4 +121,60 @@ private static boolean isConnectedNetwork(NetworkInfo networkInfo) {
     }
     return networkInfo.isConnectedOrConnecting() && networkInfo.isAvailable();
   }
+
+  /**
+   * Post-Lollipop, we can use VpnService.Builder.addDisallowedApplication to exclude Intra from its
+   * own VPN.  If for some reason we needed to get the IP address of the non-VPN resolvers, we could
+   * use ConnectivityManager.getActiveNetwork().getLinkProperties().getDnsServers().
+   *
+   * Pre-Lollipop, there is no official way to get the non-VPN DNS servers, but we need them in
+   * order to do DNS lookups on protected sockets in Go, in order to resolve DNS server names when
+   * the current server connection is broken, without disabling the VPN.  This implementation
+   * exposes the hidden version of LinkProperties that was present prior to Lollipop, in order to
+   * extract the DNS server IPs.
+   *
+   * @return The (unencrypted) DNS servers used by the underlying networks.
+   */
+  List<InetAddress> getSystemResolvers() {
+    // This list of network types is in roughly descending priority order, so that the first
+    // entries in the returned list are most likely to be the appropriate resolvers.
+    int[] networkTypes = new int[]{
+        ConnectivityManager.TYPE_ETHERNET,
+        ConnectivityManager.TYPE_WIFI,
+        ConnectivityManager.TYPE_MOBILE,
+        ConnectivityManager.TYPE_WIMAX,
+    };
+
+    List<InetAddress> resolvers = new ArrayList<>();
+    if (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP) {
+      LogWrapper.log(Log.ASSERT, LOG_TAG, "This function should never be called in L+.");
+      return resolvers;
+    }
+
+    try {
+      // LinkProperties ConnectivityManager.getLinkProperties(int type) (removed in Lollipop)
```

Comment:
Is there a way to target pre-lollipop APIs (i.e. through annotations) to avoid using reflection?

Reply:
These APIs weren't public pre-Lollipop.  Well, to be precise, the `getLinkProperties` method and the entire `LinkProperties` class were `public`, but annotated with [`@hide`](https://stackoverflow.com/questions/17035271/what-does-hide-mean-in-the-android-source-code), which [excludes them](https://commonsware.com/blog/2018/01/18/think-hard-about-hide.html) from Javadoc and `android.jar`, which is what the compiler is compiling against.  They weren't part of the "Android API".

I'm not aware of a way to access them other than reflection.  In principle, I suppose one could supplement `android.jar` with another "platform jar" exposing these additional APIs, but I've never heard of such an approach being used.

Reply:
I see. Makes sense to use reflection.